src/closure.rs

```diff
@@ -0,0 +1,136 @@
+
+use gobject_ffi;
+use translate::{ToGlibPtr, ToGlibPtrMut, Uninitialized, from_glib_none};
+use types::Type;
+use Value;
+use ToValue;
+
+glib_wrapper! {
+    pub struct Closure(Shared<gobject_ffi::GClosure>);
+
+    match fn {
+        ref => |ptr| {
+            gobject_ffi::g_closure_ref(ptr);
+            gobject_ffi::g_closure_sink(ptr);
+        },
+        unref => |ptr| gobject_ffi::g_closure_unref(ptr),
+    }
+}
+
+impl Closure {
+    pub fn new<F: Fn(&[Value]) -> Option<Value> + 'static>(callback: F) -> Self {
+        unsafe extern "C" fn marshal<F: Fn(&[Value]) -> Option<Value> + 'static>(_closure: *mut gobject_ffi::GClosure,
+            return_value: *mut gobject_ffi::GValue, n_param_values: c_uint, param_values: *mut gobject_ffi::GValue,
+            _invocation_hint: *mut c_void, marshal_data: *mut c_void)
+        {
+            let values = slice::from_raw_parts(param_values as *const _, n_param_values as usize);
+            let callback: Box<F> = Box::from_raw(marshal_data as *mut _);
+            let result = callback(values);
+            if !return_value.is_null() {
+                match result {
+                    Some(result) => *return_value = result.into_raw(),
+                    None => {
+                        let result = Value::uninitialized();
+                        *return_value = result.into_raw();
+                    },
+                }
+            }
+            mem::forget(callback);
+        }
+
+        unsafe extern "C" fn finalize<F: Fn(&[Value]) -> Option<Value> + 'static>(notify_data: *mut c_void,
+            _closure: *mut gobject_ffi::GClosure)
+        {
+            let _callback: Box<F> = Box::from_raw(notify_data as *mut _);
+            // callback is dropped here.
+        }
+
+        unsafe {
+            let size = 4 + 4 + 3 * mem::size_of::<*mut c_void>() as u32;
+            let closure = gobject_ffi::g_closure_new_simple(size, ptr::null_mut());
+            assert_ne!(closure, ptr::null_mut());
+            let callback = Box::new(callback);
+            let ptr: *mut F = Box::into_raw(callback);
+            let ptr: *mut c_void = ptr as *mut _;
+            gobject_ffi::g_closure_set_meta_marshal(closure, ptr, Some(marshal::<F>));
+            gobject_ffi::g_closure_add_finalize_notifier(closure, ptr, Some(finalize::<F>));
+            from_glib_none(closure)
+        }
+    }
+
+    pub fn invoke(&self, values: &[&ToValue]) -> Option<Value> {
+        let mut result = unsafe { Value::uninitialized() };
+        let mut values: Vec<_> = values.iter()
+            .map(|v| v.to_value())
+            .collect();
+        let gvalues = values.as_mut_ptr() as *mut _;
+        unsafe {
+            gobject_ffi::g_closure_invoke(self.to_glib_none().0 as *mut _, result.to_glib_none_mut().0,
+                values.len() as u32, gvalues, ptr::null_mut());
+        }
+        if result.type_() == Type::Invalid {
+            None
+        }
+        else {
```

Comment:
Still a style issue (so still not mandatory): could you put the `else {` on the same line as `}` please?